notebooker/web/utils.py

```diff
@@ -46,24 +46,128 @@ def get_directory_structure(starting_point: Optional[str] = None) -> Dict[str, U
+            name: original_key
+            for original_key, sub_dict in all_dirs.get(rootdir.split(os.sep)[-1], {}).items()
+            for name, value in sub_dict.items()
+            if value is None
         }
-        parent = reduce(dict.get, folders[:-1], all_dirs)
-        parent[folders[-1]] = subdir
-    return all_dirs[rootdir[start:]]
+        current_app.config["PATH_TO_CATEGORY_DICT"] = path_to_category_name
+
+    stripped = strip_extensions(all_dirs)
+    logger.info("Stripped directory structure %s", stripped)
+    return stripped.get(rootdir[start:], {})
+
+
+def strip_extensions(d):
+    def strip_extension(item):
+        """Strips .py or .ipynb extension from a given item, if present."""
+        for ext in (".py", ".ipynb"):
+            if item.endswith(ext):
+                return item[: -len(ext)]
+        return item
+
+    """
+    Recursively removes .ipynb and .py extensions from all keys and values in the dictionary.
+    """
+
+    def process_dict(sub_d):
+        """
+        Recursively processes each item in the dictionary to strip extensions from keys and values.
+        """
+        new_dict = {}
+        for k, v in sub_d.items():
+            new_key = strip_extension(k)
+            if isinstance(v, dict):
+                new_dict[new_key] = process_dict(v)  # Recursive call for sub-dictionaries
+            elif isinstance(v, str):
+                new_dict[new_key] = strip_extension(v)  # Strip extension from values if string
+            else:
+                new_dict[new_key] = v  # Copy other values directly
+        return new_dict
+
+    return process_dict(d)
+
+
+def filter_for_code_files(d):
+    """
+    Recursively filters a dictionary to retain only items that are either
+    .py or .ipynb files or directories leading to such files.
+    """
+
+    def has_code_files(sub_d):
+        """
+        Determines whether a dictionary or its nested dictionaries contain
+        any .py or .ipynb files.
+        """
+        if not isinstance(sub_d, dict):
+            return False
+        for k, v in sub_d.items():
+            if isinstance(v, dict) and has_code_files(v):
+                return True
+            if k.endswith((".py", ".ipynb")):
+                return True
+        return False
+
+    def filter_dict(sub_d):
+        """
+        Recursively filters the dictionary to retain only keys leading to .py or .ipynb files,
+        or to dictionaries that do, direct or indirectly.
+        """
+        new_dict = {}
+        for k, v in sub_d.items():
+            if isinstance(v, dict):
+                filtered_sub_d = filter_dict(v)
+                if has_code_files(filtered_sub_d):  # Retain if leads to code files
+                    new_dict[k] = filtered_sub_d
+            elif k.endswith((".py", ".ipynb")):
+                new_dict[k] = v
+        return new_dict
+
+    filtered_dict = filter_dict(d)
+    return strip_extensions(filtered_dict)
```

Comment:
How come all of this extra filtering and cleaning is required? It wasn't beforehand, and the categories do not add new files to clutter the output structure as far as I understand.